crates/runtime/src/memfd.rs

```diff
@@ -0,0 +1,519 @@
+    /// instantiation). Set by `instantiate()` and cleared by
+    /// `clear_and_remain_ready()`, and used in assertions to ensure
+    /// those methods are called properly.
+    dirty: bool,
+    /// Whether this MemFdSlot is responsible for mapping anonymous
+    /// memory (to hold the reservation while overwriting mappings
+    /// specific to this slot) in place when it is dropped. Default
+    /// on, unless the caller knows what they are doing.
+    clear_on_drop: bool,
+}
+
+impl MemFdSlot {
+    pub(crate) fn create(base_addr: *mut c_void, static_size: usize) -> Self {
+        let base = base_addr as usize;
+        MemFdSlot {
+            base,
+            static_size,
+            initial_size: 0,
+            cur_size: 0,
+            image: None,
+            dirty: false,
+            clear_on_drop: true,
+        }
+    }
+
+    /// Inform the MemFdSlot that it should *not* clear the underlying
+    /// address space when dropped. This should be used only when the
+    /// caller will clear or reuse the address space in some other
+    /// way.
+    pub(crate) unsafe fn no_clear_on_drop(&mut self) {
+        self.clear_on_drop = false;
+    }
+
+    pub(crate) fn set_heap_limit(&mut self, size_bytes: usize) -> Result<()> {
+        assert!(
+            size_bytes > self.cur_size,
+            "size_bytes = {} cur_size = {}",
+            size_bytes,
+            self.cur_size
+        );
+        // mprotect the relevant region.
+        let start = self.base + self.cur_size;
+        let len = size_bytes - self.cur_size;
+        unsafe {
+            rustix::io::mprotect(
+                start as *mut _,
+                len,
+                rustix::io::MprotectFlags::READ | rustix::io::MprotectFlags::WRITE,
+            )?;
+        }
+        self.cur_size = size_bytes;
+
+        Ok(())
+    }
+
+    pub(crate) fn instantiate(
+        &mut self,
+        initial_size_bytes: usize,
+        maybe_image: Option<&Arc<MemoryMemFd>>,
+    ) -> Result<(), InstantiationError> {
+        assert!(!self.dirty);
+
+        // Fast-path: previously instantiated with the same image, or
+        // no image but the same initial size, so the mappings are
+        // already correct; there is no need to mmap anything. Given
+        // that we asserted not-dirty above, any dirty pages will have
+        // already been thrown away by madvise() during the previous
+        // termination.  The `clear_and_remain_ready()` path also
+        // mprotects memory above the initial heap size back to
+        // PROT_NONE, so we don't need to do that here.
+        if (self.image.is_none()
+            && maybe_image.is_none()
+            && self.initial_size == initial_size_bytes)
+            || (self.image.is_some()
+                && maybe_image.is_some()
+                && self.image.as_ref().unwrap().fd.as_file().as_raw_fd()
+                    == maybe_image.as_ref().unwrap().fd.as_file().as_raw_fd())
```

Comment:
Random thought that just occured to me, since we're doing fd equality matching here do we actually still need a compiled module id? I originally thought that was going to be used for this but now that I look at this again I see it not being used. I forget though if the compiled module id is actually used for anything else at this point in time though.

Reply:
It's used in the followup affinity PR (#3738) as otherwise there's an ID-reuse problem... strictly speaking that's just a heuristic so we could use fds there too and accept a false affinity sometimes. IMHO they're pretty cheap to maintain so maybe it's ok...?

Reply:
Ah right, and nah yeah let's leave them in. Only worthwhile if they could be removed entirely, but using for affinity makes sense!